### DISCUSSION TITLE
tracing-aka-logging  help

### DISCUSSION THREAD

**Discussion by ximengyi** [tokio-rs/axum#689] [Q&A] [answered]
I am learning to use axum
I cloned axum and went directly to /axum/examples/tracing-aka-logging to run cargo run
I can see:
```
2022-01-04T08:32:52.673597Z DEBUG example_tracing_aka_logging: listening on 127.0.0.1:3000
```
But I copied this example into my own project, but I can’t print this sentence. My code is exactly the same as the code in tracing-aka-logging.
But 
```
tracing::debug!("listening on {}", addr);  //This code seems to be invalid.
```
This code cannot be printed to the console normally, I am confused ，Can you help me？

Below is the complete code and dependencies there is only one main.rs file

```rust
use axum::{
    body::Bytes,
    http::{HeaderMap, Request},
    response::{Html, Response},
    routing::get,
    Router,
};
use std::{net::SocketAddr, time::Duration};
use tower_http::{classify::ServerErrorsFailureClass, trace::TraceLayer};
use tracing::Span;

#[tokio::main]
async fn main() {
    // Set the RUST_LOG, if it hasn't been explicitly defined
    if std::env::var_os("RUST_LOG").is_none() {
        std::env::set_var(
            "RUST_LOG",
            "example_tracing_aka_logging=debug,tower_http=debug",
        )
    }
    tracing_subscriber::fmt::init();

    // build our application with a route
    let app = Router::new()
        .route("/", get(handler))
        // `TraceLayer` is provided by tower-http so you have to add that as a dependency.
        // It provides good defaults but is also very customizable.
        //
        // See https://docs.rs/tower-http/0.1.1/tower_http/trace/index.html for more details.
        .layer(TraceLayer::new_for_http())
        // If you want to customize the behavior using closures here is how
        //
        // This is just for demonstration, you don't need to add this middleware twice
        .layer(
            TraceLayer::new_for_http()
                .on_request(|_request: &Request<_>, _span: &Span| {
                    // ...
                })
                .on_response(|_response: &Response, _latency: Duration, _span: &Span| {
                    // ...
                })
                .on_body_chunk(|_chunk: &Bytes, _latency: Duration, _span: &Span| {
                    // ..
                })
                .on_eos(
                    |_trailers: Option<&HeaderMap>, _stream_duration: Duration, _span: &Span| {
                        // ...
                    },
                )
                .on_failure(
                    |_error: ServerErrorsFailureClass, _latency: Duration, _span: &Span| {
                        // ...
                    },
                ),
        );

    // run it
    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    tracing::debug!("listening on {}", addr);
    tracing::info!("listening on {}", addr);
    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
}

async fn handler() -> Html<&'static str> {
     tracing::debug!("hello world");
    Html("<h1>Hello, World!</h1>")
}
``` 
```
[package]
name = "glimmer"
version = "0.1.0"
edition = "2021"
publish = false


[dependencies]
axum = { version = "0.4" }
tokio = { version = "1.0", features = ["full"] }
tracing = "0.1"
tracing-subscriber = { version="0.3", features = ["env-filter"] }
tower-http = { version = "0.2", features = ["trace"] }
```

**Comment by davidpdrsn** [comment, marked as answer]
Since your crate has a different name you need to update the log filter:

```diff
std::env::set_var(
    "RUST_LOG",
-   "example_tracing_aka_logging=debug,tower_http=debug",
+   "glimmer=debug,tower_http=debug",
)
```

**Reply by ximengyi** [reply]
It's okay thank you very much

**Comment by wemgl** [comment]
Hey @davidpdrsn, I came across this issue after running into the same problem. The solution you provided works, but I'd like to know where in the documentation I can find a description of an acceptable log filter configuration. I've already looked through https://github.com/tokio-rs/tower-http repo, and although the examples in that repo do use the crate's name in the log filter it didn't register to me that it **had** to be the crate's name, so I changed it to something else and saw that the log levels didn't change as I expected.

**Reply by davidpdrsn** [reply]
You'll wanna look at tracing's docs.

**Reply by wemgl** [reply]
Thanks, I was able to find it here: https://docs.rs/tracing-subscriber/0.2.12/tracing_subscriber/fmt/index.html#filtering-events-with-environment-variables

**Reply by 1oglop1** [reply]
@davidpdrsn This is the most confusing thing I find as a beginner. The documentation is scattered across way too many places and it takes a lot of time to get something done or learn. 
I'm comparing (Idk if it's a fair comparison) `axum` to `Flask` or `Starlette` where docs make more sense to me.

**Reply by davidpdrsn** [reply]
Do you have any suggestions for how to improve the situation? Keeping in mind that we can't simply document _everything_ as part of axum.

**Reply by 1oglop1** [reply]
I think that better interlinking between projects would help a lot. If docs suggested checking out related projects, eg. `for this <concept> visit <X project>`
